lib/merge.js

```diff
@@ -0,0 +1,140 @@
+function flatten(array) {
+  if (array.flat) {
+    return array.flat();
+  }
+  return [].concat(...array);
+}
+
+// https://heycam.github.io/webidl/#own-exposure-set
+function getOwnExposureSet(node) {
+  const exposedAttr = node.extAttrs.find((a) => a.name === "Exposed");
+  if (!exposedAttr) {
+    return null;
+  }
+  const exposure = new Set();
+  const { type, value } = exposedAttr.rhs;
+  if (type === "identifier") {
+    exposure.add(value);
+  } else if (type === "identifier-list") {
+    for (const ident of value) {
+      exposure.add(ident.value);
+    }
+  }
+  return exposure;
+}
+
+/**
+ * @param {Set?} a a Set or null
+ * @param {Set?} b a Set or null
+ * @return {Set?} a new intersected set, one of the original sets, or null
+ */
+function intersectNullable(a, b) {
+  if (a && b) {
+    const intersection = new Set();
+    for (const v of a.values()) {
+      if (b.has(v)) {
+        intersection.add(v);
+      }
+    }
+    return intersection;
+  }
+  return a || b;
+}
+
+/**
+ * @param {Set?} a a Set or null
+ * @param {Set?} b a Set or null
+ * @return true if a and b have the same values, or both are null
+ */
+function equalsNullable(a, b) {
+  if (a && b) {
+    if (a.size !== b.size) {
+      return false;
+    }
+    for (const v of a.values()) {
+      if (!b.has(v)) {
+        return false;
+      }
+    }
+  }
+  return a === b;
+}
+
+/**
+ * @param {Container} target definition to copy members to
+ * @param {Container} source definition to copy members from
+ */
+function copyMembers(target, source) {
+  const targetExposure = getOwnExposureSet(target);
+  const parentExposure = intersectNullable(
+    targetExposure,
+    getOwnExposureSet(source)
+  );
+  // TODO: extended attributes
+  for (const orig of source.members) {
+    const origExposure = getOwnExposureSet(orig);
+    const copyExposure = intersectNullable(origExposure, parentExposure);
+
+    // Make a copy of the member with the same prototype and own properties.
+    const copy = Object.create(
+      Object.getPrototypeOf(orig),
+      Object.getOwnPropertyDescriptors(orig)
+    );
+
+    if (!equalsNullable(targetExposure, copyExposure)) {
+      let value = Array.from(copyExposure.values()).join(",");
+      if (copyExposure.size !== 1) {
+        value = `(${value})`;
+      }
+      copy.extAttrs = ExtendedAttributes.parse(
+        new Tokeniser(` [Exposed=${value}] `)
+      );
+    }
+
+    target.members.push(copy);
+  }
+}
+
+/**
+ * @param {*[]} ast AST or array of ASTs
+ * @return {*[]}
+ */
+export function merge(ast) {
+  const dfns = new Map();
+  const partials = [];
+  const includes = [];
+
+  for (const dfn of flatten(ast)) {
+    if (dfn.partial) {
+      partials.push(dfn);
+    } else if (dfn.type === "includes") {
+      includes.push(dfn);
+    } else if (dfn.name) {
+      dfns.set(dfn.name, dfn);
+    } else {
+      throw new Error(`definition with no name`);
+    }
+  }
+
+  // merge partials (including partial mixins)
```

Comment:
The idea here is to first merge partials and then merge mixins, performing basically the same steps each time. But this will have the downside of copying members and merging extended attributes twice in the case of a partial interface mixin. (The mixin step isn't here yet.)

Performance isn't much of a concern, but I wonder if this is overgeneralized, in particular it becomes hard to judge whether https://heycam.github.io/webidl/#dfn-exposure-set is followed.

The alternative would be to first collect all includes statements and then to copy members directly from their source to their final destination.

Reply:
>in particular it becomes hard to judge whether https://heycam.github.io/webidl/#dfn-exposure-set is followed.

Not sure I follow this part, in what situation would it be problematic?

Reply:
I mean if the implementation isn't similar to the algorithm in the spec, it's harder to know if it get the same result in all cases. I think the two-step approach would be equivalent, but I'm not certain. The rules about whether the extended attributes of the member have to be a subset of the container's or if the intersection is used differ at least, so a unified implementation would allow for more invalid stuff.